### PR TITLE
Bindgen as a workspace

### PR DESCRIPTION
**Pull request by pvdrz**
This reorganizes the repo into a virtual workspace. The main change is that `bindgen` as a build-script library and `bindgen` as a command line utility have been split into the `bindgen` and `bindgen-cli` crates respectively.

This means that `bindgen` no longer depends on `clap`. Transforming it's `cargo tree` from this:
```
bindgen v0.60.1 (/home/christian/Workspace/bindgen/rust-bindgen)
├── bitflags v1.3.2
├── cexpr v0.6.0
│   └── nom v7.0.0
│       ├── memchr v2.5.0
│       └── minimal-lexical v0.1.4
│       [build-dependencies]
│       └── version_check v0.9.3
├── clang-sys v1.3.3
│   ├── glob v0.3.0
│   ├── libc v0.2.126
│   └── libloading v0.7.0
│       └── cfg-if v1.0.0
│   [build-dependencies]
│   └── glob v0.3.0
├── clap v3.2.12
│   ├── atty v0.2.14
│   │   └── libc v0.2.126
│   ├── bitflags v1.3.2
│   ├── clap_lex v0.2.4
│   │   └── os_str_bytes v6.2.0
│   ├── indexmap v1.9.1
│   │   └── hashbrown v0.12.2
│   │   [build-dependencies]
│   │   └── autocfg v1.1.0
│   ├── strsim v0.10.0
│   ├── termcolor v1.1.3
│   └── textwrap v0.15.0
├── env_logger v0.9.0
│   ├── atty v0.2.14 (*)
│   ├── humantime v2.1.0
│   ├── log v0.4.14
│   │   └── cfg-if v1.0.0
│   ├── regex v1.5.5
│   │   ├── aho-corasick v0.7.18
│   │   │   └── memchr v2.5.0
│   │   ├── memchr v2.5.0
│   │   └── regex-syntax v0.6.25
│   └── termcolor v1.1.3
├── lazy_static v1.4.0
├── lazycell v1.3.0
├── log v0.4.14 (*)
├── peeking_take_while v0.1.2
├── proc-macro2 v1.0.43
│   └── unicode-ident v1.0.3
├── quote v1.0.9
│   └── proc-macro2 v1.0.43 (*)
├── regex v1.5.5 (*)
├── rustc-hash v1.1.0
├── shlex v1.0.0
├── syn v1.0.99
│   ├── proc-macro2 v1.0.43 (*)
│   ├── quote v1.0.9 (*)
│   └── unicode-ident v1.0.3
└── which v4.2.2
    ├── either v1.6.1
    └── libc v0.2.126
[dev-dependencies]
├── clap v3.2.12 (*)
├── diff v0.1.12
├── shlex v1.0.0
└── tempfile v3.2.0
    ├── cfg-if v1.0.0
    ├── libc v0.2.126
    ├── rand v0.8.4
    │   ├── libc v0.2.126
    │   ├── rand_chacha v0.3.1
    │   │   ├── ppv-lite86 v0.2.10
    │   │   └── rand_core v0.6.3
    │   │       └── getrandom v0.2.3
    │   │           ├── cfg-if v1.0.0
    │   │           └── libc v0.2.126
    │   └── rand_core v0.6.3 (*)
    └── remove_dir_all v0.5.3
```
to this:
```
bindgen v0.60.1 (/home/christian/Workspace/bindgen/rust-bindgen/bindgen)
├── bitflags v1.3.2
├── cexpr v0.6.0
│   └── nom v7.0.0
│       ├── memchr v2.5.0
│       └── minimal-lexical v0.1.4
│       [build-dependencies]
│       └── version_check v0.9.3
├── clang-sys v1.3.3
│   ├── glob v0.3.0
│   ├── libc v0.2.126
│   └── libloading v0.7.0
│       └── cfg-if v1.0.0
│   [build-dependencies]
│   └── glob v0.3.0
├── lazy_static v1.4.0
├── lazycell v1.3.0
├── log v0.4.14
│   └── cfg-if v1.0.0
├── peeking_take_while v0.1.2
├── proc-macro2 v1.0.43
│   └── unicode-ident v1.0.3
├── quote v1.0.9
│   └── proc-macro2 v1.0.43 (*)
├── regex v1.5.5
│   └── regex-syntax v0.6.25
├── rustc-hash v1.1.0
├── shlex v1.0.0
├── syn v1.0.99
│   ├── proc-macro2 v1.0.43 (*)
│   ├── quote v1.0.9 (*)
│   └── unicode-ident v1.0.3
└── which v4.2.2
    ├── either v1.6.1
    └── libc v0.2.126
```
There are some questions that I'd like to resolve before moving forward with this:
- [ ] Is CI working correctly? There are many test crates like `quickcheck` and `expect` but I'm not sure if my changes to the ci scripts are actually running everything they should.
- [x] How do we distribute `bindgen-cli`? RIght now you can test `cargo install` running the command
   ```bash
   cargo install --git=https://github.com/ferrous-systems/rust-bindgen --branch=bindgen-as-workspace bindgen-cli
   ```
   Which installs a `bindgen-cli` binary instead of just `bindgen`. I haven't found out how to change the name of the binary back to just `bindgen`.

### CODE REVIEW

**Comment by emilio** [comment]
This looks great, though it'd be great to not change the name of the CLI tool. It seems something like this would do?

```diff
diff --git a/bindgen-cli/Cargo.toml b/bindgen-cli/Cargo.toml
index 2aef9b5b..ebeceacc 100644
--- a/bindgen-cli/Cargo.toml
+++ b/bindgen-cli/Cargo.toml
@@ -21,6 +21,10 @@ include = [
   "build.rs",
 ]
 
+[[bin]]
+path = "src/main.rs"
+name = "bindgen"
+
 [badges]
 travis-ci = { repository = "rust-lang/rust-bindgen" }
```

Also, we need to update the docs that mention `cargo install` and so on.

File: bindgen-tests/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+[package]
+authors = [
+  "The rust-bindgen project contributors",
+]
+description = "Automatically generates Rust FFI bindings to C and C++ libraries."
```
Comment (emilio) [inline bindgen-tests/Cargo.toml:5]: This should probably be removed, and we should not publish this crate.

File: bindgen-cli/build.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        // We may not have headers directory after packaging.
+        Err(..) => return,
+    };
+
+    let entries =
+        headers.map(|result| result.expect("Couldn't read header file"));
+
+    println!("cargo:rerun-if-changed=tests/headers");
+
+    for entry in entries {
```
Comment (emilio) [inline bindgen-cli/build.rs:26]: Do you need this? If so, why do we need the `bindgen-tests` crate?

File: bindgen-integration/build.rs [hunk 3]
```diff
@@ -123,13 +123,9 @@ impl ParseCallbacks for MacroCallback {
     // Test the "custom derives" capability by adding `PartialEq` to the `Test` struct.
     fn add_derives(&self, name: &str) -> Vec<String> {
         if name == "Test" {
-            vec![
-                "PartialEq".into(),
-            ]
+            vec!["PartialEq".into()]
```
Comment (emilio) [inline bindgen-integration/build.rs:126]: These changes seem drive-by?

File: bindgen-tests/tests/expectations/struct_with_anon_struct_array_float.rs [hunk 4]
```diff
@@ -0,0 +1 @@
+
```
Comment (emilio) [inline bindgen-tests/tests/expectations/struct_with_anon_struct_array_float.rs:1]: Huh?

File: ci/test.sh [hunk 5]
```diff
@@ -126,11 +126,11 @@ fi
 CARGO_ARGS=`get_cargo_args`
 
 # Ensure we build without warnings
-cargo rustc --lib $CARGO_ARGS -- -Dwarnings
+(cd bindgen-tests && cargo rustc --lib $CARGO_ARGS -- -Dwarnings)
```
Comment (emilio) [inline ci/test.sh:129]: This was testing that bindgen built without warnings, but is no longer doing that is it?

**Review by emilio — CHANGES_REQUESTED** [review]
Other thoughts:

 * Maybe we can flatten the directory structure a bit by using `Cargo.toml` magic (e.g. don't require `bindgen/src`, `bindgen-cli/src` and so on).
 * It seems some of the test scripts need a bit of work.
 * The naming of the cli crate I believe should probably be kept, I think what I posted earlier should work but lmk if it doesn't for any reason.

Other than that, this looks pretty good!

File: bindgen-integration/build.rs (hunk 3 above)
Comment (pvdrz) [reply bindgen-integration/build.rs:126]: I think they happened because cargo didn't consider `bindgen-integration` a proper crate before so `rustfmt` never ran over it.

File: bindgen-tests/tests/expectations/struct_with_anon_struct_array_float.rs (hunk 4 above)
Comment (pvdrz) [reply bindgen-tests/tests/expectations/struct_with_anon_struct_array_float.rs:1]: this seems to be a merge conflict casualty

Comment (pvdrz) [reply bindgen-tests/tests/expectations/struct_with_anon_struct_array_float.rs:1]: It is not :upside_down_face: the file is just there since 2017: https://github.com/rust-lang/rust-bindgen/blob/master/tests/expectations/struct_with_anon_struct_array_float.rs

File: bindgen-cli/build.rs (hunk 2 above)
Comment (pvdrz) [reply bindgen-cli/build.rs:26]: the crate exists so `bindgen` does not depend on `bindgen-cli` for the tests.

**Comment by pvdrz** [comment]
@emilio the `[[bin]]` section on the manifest worked. You still have to install `bindgen-cli` but get a `bindgen` executable.

File: bindgen-cli/build.rs (hunk 2 above)
Comment (emilio) [reply bindgen-cli/build.rs:26]: I meant the build script, is this really doing something? This seems to be trying to read `$CARGO_MANIFEST_DIR/tests/headers`, which no longer exists... So can't we just remove it?

Comment (pvdrz) [reply bindgen-cli/build.rs:26]: isn't `$CARGO_MANIFEST_DIR` just `bindgen-tests/`? I printed earch `DirEntry` in the loop and `cargo build -vv -p bindgen-tests` is printing stuff so it seems the path is the right one: `bindgen-tests/tests/headers`

Comment (emilio) [reply bindgen-cli/build.rs:26]: Are you looking at `bindgen-cli/build.rs` or `bindgen-tests/build.rs`?

Comment (emilio) [reply bindgen-cli/build.rs:26]: (I don't understand why both are needed)

Comment (pvdrz) [reply bindgen-cli/build.rs:26]: ohhhh, that actually explains a lot, my bad :sweat_smile: . Yeah It seems I forgot to remove it

**Review by emilio — APPROVED** [review]
Let's get this merged sooner rather than later to avoid all the conflicts.

Can you squash your commits before landing please? Thank you for doing this!

File: bindgen-tests/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+edition = "2018"
+version = "0.1.0"
+
+include = [
+  "Cargo.toml",
+  "build.rs",
+]
+
+[dev-dependencies]
+bindgen = { version = "=0.60.1", path = "../bindgen" }
```
Comment (emilio) [inline bindgen-tests/Cargo.toml:12]: Do we really need the explicit version? That means that bumping bindgen would also need bumping this which seems unfortunate.

File: bindgen-cli/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+[[bin]]
+path = "main.rs"
+name = "bindgen"
+
+[badges]
+travis-ci = { repository = "rust-lang/rust-bindgen" }
+
+[dependencies]
+bindgen = { version = "=0.60.1", path = "../bindgen" }
```
Comment (emilio) [inline bindgen-cli/Cargo.toml:32]: Same comment about this, though I guess here it can be justified.

**Review by emilio — COMMENTED** [review]
A couple last minute comments.

**Comment by pvdrz** [comment]
Hi everyone! If you are reading this message because this PR broke yours, I'm sorry :grimacing:. For most of the cases it is enough to rebase against master as there won't be any conflicts, just files being movede here and there

File: bindgen-tests/tests/expectations/struct_with_anon_struct_array_float.rs (hunk 4 above)
Comment (pvdrz) [reply bindgen-tests/tests/expectations/struct_with_anon_struct_array_float.rs:1]: Apparently the file was created here and it was empty from the start. I'll remove it in an upcoming: https://github.com/fitzgen/rust-bindgen/commit/b4895d96698504dc28dfa64c4c4e4c3f3714a3c8
